File: Algorithm/src/PeakFinder.java

public class PeakFinder {
	
	// A binary search based function that returns index of a peak 
    // element 
	//n = array's length
    
	static int findPeakUtil(int arr[], int front, int end, int n) 
    { 
        // Find index of middle element 
        int mid = front + (end - front)/2;  /* (low + high)/2   avoid overflow */ 
  
        // Compare middle element with its neighbours (if neighbours 
        // exist) 
        
        if ((mid == 0 || arr[mid-1] <= arr[mid]) && (mid == n-1 || 
             arr[mid+1] <= arr[mid])) 
            return mid; 
        
        
        //如果 mid-1 > mid , 回傳從左邊 front 到 mid-1 進行比較
        else if (mid > 0 && arr[mid-1] > arr[mid]) 
        	return findPeakUtil(arr, front, (mid -1), n); 
        
        
        //如果 mid+1> mid, 回傳右邊 mid+1 到 end 進行比較
        else return findPeakUtil(arr, (mid + 1), end, n); 
    } 
  
	
    static int findPeak(int arr[], int n)    
    { 
        return findPeakUtil(arr, 0, n-1, n); 
    } 
  
    // Driver method 
    public static void main (String[] args) 
    { 
        int arr[] = {1, 3, 20, 4, 1, 0}; 
        int n = arr.length; 
        System.out.println("Index of a peak point is " + 
                            findPeak(arr, n)); 
    } 
}

File: Algorithm/src/Peak/peakfinder.py

def untilfindpeak(array,front,end,n):
  
  mid = (front + end)/2
  mid = int(mid)
  
  if (mid==0) or (mid == n-1):
    return mid
  if array[mid]>=array[mid-1] and array[mid] >= array[mid+1]:
    return mid 
  elif mid>0 and array[mid-1]>array[mid]:
    return untilfindpeak(array,front,mid-1,n)
  else:
    return untilfindpeak(array,mid+1,end,n)

def peakfinder(array,n):
  return untilfindpeak(array,0,n-1,n)

